File: src/core/caps/kernel_cap/btf.rs
// Copyright (C) 2026 rezky_nightky
// SPDX-License-Identifier: GPL-3.0-only

//! BTF (BPF Type Format) probe.
//!
//! Uses: kernel_bool()

use crate::core::capability::Capability;
use crate::core::caps::kernel_cap::kernel_bool;
use crate::core::evidence::Evidence;

pub struct KernelBtfStatus;

impl Capability for KernelBtfStatus {
    fn id(&self) -> &'static str {
        "kernel.btf.module"
    }
    fn label(&self) -> &'static str {
        "BTF Status (module)"
    }
    fn probe(&self) -> Evidence {
        kernel_bool(self.id(), "btf.available")
    }
}

File: kernel/probes/btf.c
// SPDX-License-Identifier: GPL-2.0-only
// Copyright (C) 2026 rezky_nightky

//! BTF (BPF Type Format) availability probe.
//!
//! Discovered facts:
//!   btf.available   — "yes" | "no"
//!   btf.vmlinux     — "available" | "unavailable"
//!
//! Source: CONFIG_DEBUG_INFO_BTF compile-time + runtime btf_vmlinux check.

#include <linux/kernel.h>
#include <linux/string.h>
#include "zenvecha.h"

static char available_buf[8] = "yes";
static char vmlinux_buf[16] = "available";

static const struct capability_descriptor descriptors[] = {
	{ .key = "btf.available",  .value = available_buf },
	{ .key = "btf.vmlinux",    .value = vmlinux_buf   },
};

const struct capability_descriptor *btf_probe_discover(void)
{
#ifdef CONFIG_DEBUG_INFO_BTF
	snprintf(available_buf, sizeof(available_buf), "yes");
	snprintf(vmlinux_buf, sizeof(vmlinux_buf), "available");
#else
	snprintf(available_buf, sizeof(available_buf), "no");
	snprintf(vmlinux_buf, sizeof(vmlinux_buf), "unavailable");
#endif

	return descriptors;
}

size_t btf_probe_count(void)
{
	return ARRAY_SIZE(descriptors);
}
